New API for GDNative

vnen opened the pull request:
This removes the old functions that were wrappers for builtin functions and instead provide the new discovery API to get function pointers for those.

Also removes the `print*()` functions since those are utility functions that also can be queried.

vnen committed:
GDNative: New core API

This API now uses the discovery functions present in Variant instead of
wrapping every built-in function. Users now need to query for function
pointers and use those.

vnen committed:
GDNative: Remove print functions

Those are now utilities so the function pointer can be fetched when
needed.

akien-mga commented on modules/gdnative/gdnative/packed_arrays.cpp:
```diff
@@ -30,882 +30,101 @@
+static_assert(sizeof(godot_packed_float32_array) == sizeof(PackedFloat32Array), "PackedFloat32Array size mismatch");
+static_assert(sizeof(godot_packed_float64_array) == sizeof(PackedFloat64Array), "PackedFloat64Array size mismatch");
+static_assert(sizeof(godot_packed_string_array) == sizeof(PackedStringArray), "PackedStringArray size mismatch");
+static_assert(sizeof(godot_packed_vector2_array) == sizeof(PackedVector2Array), "PackedVector2Array size mismatch");
+static_assert(sizeof(godot_packed_vector2i_array) == sizeof(Vector<Vector2i>), "Vector<Vector2i> size mismatch");
+static_assert(sizeof(godot_packed_vector3_array) == sizeof(PackedVector3Array), "PackedVector3Array size mismatch");
```
Vector3i seems missing.

vnen replied:
There's not really a PackedVector3iArray. I only left the Vector2i because the text server extension is using it. I guess I can add a Vector3i for completeness.

akien-mga commented on modules/gdnative/gdnative/packed_arrays.cpp:
```diff
@@ -30,882 +30,101 @@
 #include "core/variant/variant.h"
 
-#include "core/math/color.h"
 #include "core/math/vector2.h"
```
It's weird that you'd still need `core/math/vector2.h` while the other ones are no longer used. What's the difference?

vnen replied:
Because of the Vector2i array. For the others I'm using the Packed* name which is defined in Variant, so I don't need the specific headers. Well, now that I think of it Variant also includes Vector2 so I could probably remove this anyway.

akien-mga commented on modules/gdnative/gdnative/callable.cpp:
```diff
@@ -30,223 +30,24 @@
 
-#include "core/io/resource.h"
 #include "core/variant/callable.h"
 #include "core/variant/variant.h"
```
Variant might be unused now, no?

vnen replied:
The problem is that Callable needs Variant (it's only forward-declared in the header), so I need a full definition otherwise it can't compile.

akien-mga commented on modules/gdnative/gdnative/color.cpp:
```diff
@@ -31,178 +31,17 @@
-	*((Color *)&dest) = self->lerp(*b, p_t);
-	return dest;
-}
+#include <gdnative/gdnative.h>
```
You didn't add this for other classes, is this unused?

vnen replied:
Yeah, copy-paste mistake. I double-checked it but this one still slipped through.

akien-mga commented on modules/gdnative/gdnative/rid.cpp:
```diff
@@ -30,45 +30,17 @@
 #include "gdnative/rid.h"
 
-#include "core/io/resource.h"
+#include "core/os/memory.h"
```
Unused?

vnen replied:
I use this for the `memnew_placement` macro.

akien-mga commented on modules/gdnative/gdnative/signal.cpp:
```diff
@@ -0,0 +1,53 @@
+#include "gdnative/signal.h"
+
+#include "core/variant/callable.h"
+#include "core/variant/variant.h"
```
Unused?

vnen replied:
Same thing as the Callable one. Although I think the Signal definition should be on its own header, and not in `callable.h`.

akien-mga commented on modules/gdnative/gdnative/string.cpp:
```diff
@@ -30,57 +30,15 @@
 extern "C" {
 #endif
 
-static_assert(sizeof(godot_char16_string) == sizeof(Char16String), "Char16String size mismatch");
-static_assert(sizeof(godot_char_string) == sizeof(CharString), "CharString size mismatch");
```
Are you sure that `CharString` and `Char16String` don't need to bound anymore? (I don't know, just raising it for discussion). CC @bruvzg

vnen replied:
I don't see how having those as types can be useful. AFAIK Godot doesn't deal with those directly in the API, but only with Strings. External code can use their own char types, since there are functions to make a `godot_string` out of those.

akien-mga commented on modules/gdnative/gdnative/string_name.cpp:
```diff
@@ -31,54 +31,27 @@
 	StringName *dest = (StringName *)r_dest;
-	const String *name = (const String *)p_name;
-	memnew_placement(dest, StringName(*name));
+	memnew_placement(dest, String);
```
Shouldn't this be StringName?

vnen replied:
Good catch, I'll fix it.

akien-mga commented on modules/gdnative/include/gdnative/basis.h:
```diff
@@ -46,88 +46,17 @@ typedef struct {
 } godot_basis;
 #endif
 
-// reduce extern "C" nesting for VS2013
```
We don't support VS 2013 anymore, so if all compilers we support are fine with it, we could drop the close / reopen around includes. Doesn't hurt though.

vnen replied:
I left it mostly because it was already there. I guess we can drop it now.

akien-mga commented on modules/gdnative/include/gdnative/string_name.h:
```diff
@@ -58,19 +58,12 @@ typedef struct {
+void GDAPI godot_string_name_new_copy(godot_string_name *r_dest, const godot_string_name *p_src);
 void GDAPI godot_string_name_destroy(godot_string_name *p_self);
 
+void GDAPI godot_string_name_new_with_latin1_chars(godot_string_name *r_dest, const char *p_contents);
```
That seems to be a pretty specific survivor to the great culling :D

vnen replied:
Yeah, I think this is still useful. The only other way to make a `StringName` out of `const char *` is to make a `String` first, which seems wasteful when you want to, say, use it as a member name to get the value. I left `latin1` since that's the default constructor and usually identifiers fall into this range.

akien-mga reviewed (commented):
Amazing cleanup. I like PRs with this kind of additions/removals ratio :)

akien-mga commented:
Windows build seems to fail with:
```
D:\a\godot\godot\modules\gdnative/gdnative.h(40): fatal error C1083: Cannot open include file: 'gdnative_api_struct.gen.h': No such file or directory
```

akien-mga commented:
Aside from `gdnative/variant.cpp`, all `.cpp` are now pretty short and simple (new + sometimes destroy), there could be a case for grouping them all in some `gdnative/variant_bind.cpp` or similar à la `core`... but that's also arguably a bad habit that we have to make giganormous `.cpp` files of related components, so it might be fine to keep as is ;)

vnen commented:
> Windows build seems to fail with:
> 
> ```
> D:\a\godot\godot\modules\gdnative/gdnative.h(40): fatal error C1083: Cannot open include file: 'gdnative_api_struct.gen.h': No such file or directory
> ```

I'm not sure why this is happening. Is it failing to generate this file? Reminds of the old issue with parallel builds on Windows, but I think that was fixed already.

vnen commented:
Regarding the build issue, I've exhausted my ideas. It seems the WebRTC module needs GDNative but for some reason it doesn't trigger the generation of the API struct header. On my machine it works. I wonder if it's something related to the CI cache.

akien-mga commented:
> Regarding the build issue, I've exhausted my ideas. It seems the WebRTC module needs GDNative but for some reason it doesn't trigger the generation of the API struct header. On my machine it works. I wonder if it's something related to the CI cache.

It does call
```
build_gdnative_api_struct(["modules\gdnative\include\gdnative_api_struct.gen.h", "modules\gdnative\gdnative_api_struct.gen.cpp"], ["modules\gdnative\gdnative_api.json"])
```
but it seems that there's a race condition, the file is not generated in time to be used by the webrtc gdnative code.

I'd posit that the race condition is likely a pre-existing bug (the dependency from `modules/webrtc/` on the generation of `modules\gdnative\include\gdnative_api_struct.gen.h` is not registered properly) which is exposed by this PR which modifies only GDNative and therefore can reuse most of the cache. So the build can very quickly start both the file generation and the webrtc gdnative builds in parallel, which fails (while it might work on clean builds as the file generation would likely happen much earlier).

CC @Faless

Faless commented:
@akien-mga is there any way I can tell scons that one module depends on the other so they are built sequentially?

akien-mga commented:
@Faless No easy way that I can think of, at least with our current setup. The build order likely comes from how the `SCsub` files are called in `modules/SCsub`, so we might need to hardcode some hacks there to make sure that GDNative comes last - I think the build order is reversed compared to the calling order for the `SConscript` (see e.g. how the folders are listed in the main `SConscript`, `core` is listed first and built last.

Current order for modules seems to be alphabetical.

akien-mga commented:
Testing this hack, let's see: https://github.com/akien-mga/godot/commit/5b0e522fbeaea0f7c2f442344c0ddec8ea87f76d

vnen commented:
I updated this to add a version of the discovery API to take a regular `const char *` instead of a `godot_string_name *`. For the C++ bindings (and probably others too) it's pretty useful not need to create a StringName for each call.

I've also added a few missing methods to the builtin methods discovery (to get argument count, type and name, and the return type, besides checking if they are const or vararg).

vnen commented:
But it's still failing to build. I will try to take a look in the tests.

vnen commented:
I think I fixed this now. Let's see if the CI finishes correctly.

akien-mga commented:
Thanks!
